### DISCUSSION TITLE
Branches setup

### DISCUSSION THREAD

**Discussion by Andre601** [PlaceholderAPI/PlaceholderAPI#743] [(3) Ideas]
### Type
> What kind of request is this? (Multiple selections possible)

- [ ] New function for PlaceholderAPI.  
A new function that developers could use.
- [ ] Change to code (Internal).  
Changes to code that won't affect the end-user.
- [ ] Change to code (External).  
Changes to code that will affect the end-user (breaks stuff).
- [x] Other: Repository 

### Info
> What is the change?  
> Please provide as much information (including links, images and code-snippeds) as possible.

This suggestion was originally made by @LemmoTresto 

The idea is to have the repository setup in a way, to have multiple branches for various purposes:

- `master` branch would be the main branch for releases. This should be a protected branch to only allow pull requests towards it.
- `development` would be the branch for actual updates (dev builds) of PlaceholderAPI. This should be a protected branch to only allow Pull requests towards it. Once enough changes have been made to make a release would this branch be merged into master using a PR, before creating a new Release.
- `docs/*` would be for the documentation of PlaceholderAPI. For example woul `docs/wiki` contain the up-to-date changes of the wiki and `docs/javadoc` the recent Javadoc. Those branches could be merged into the main or development branch once enough changes have been made but this doesn't have to be a requirement I would assume.

Additionally would it be recommendet to have new branch names follow specific naming conventions for easier understanding.

- Branches that contain fixes towards an issue should have the name `fix/issue-<id>` (if GitHub issue exist, otherwise `fix/<what fix?>`)
- Branches that contain new features or changes to existing features, that don't fix bugs should be named `feature/issue-<id>` (If GitHub issue exists otherwise `feature/<what changes?>`)

The branch naming is only a suggestion tho.

**Comment by LemmoTresto** [comment]
Actually there should be an issue made for everything you want to write code for.

Here's a copy from the discord:

**Branch policy**
We have 2 restricted branches we should **only** be **PRed** to.
These branches are `master` and `develop`.

- `master` is for releases only and should **only** be **PRed** from the `develop` branch.
- `develop` is where all **feature** and **bugfix** branches are **PRed** to. This branch will be used to get all features and bugfixes for the next release together in one branch.

Naming of **feature** or **bugfix** branches:
- `feature/%ISSUE_ID%-%ISSUE_TITLE%`
- `bugfix/%ISSUE_ID%-%ISSUE_TITLE%`

Make sure `%ISSUE_TITLE%` is lowercase and spaces replaced with dashes otherwise git won't accept your branch.

**Comment by Andre601** [comment]
Just want to give some info here since stuff has changed:

- The Contributing file mentions the branch setup with `development`, `master` and `docs/wiki` (Which one should be targeted for what)
- Restriction is (still) not set on the branches sadly. But at least I myself try to stick with the branch system as good as I can.
